### PR TITLE
[SPARK-40014][SQL] Support cast of decimals to ANSI intervals

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to support casts of decimals to ANSI intervals, and follow the SQL standard:
173663908-71945980-5638-4b46-9020-4d2e4badef0c

### Why are the changes needed?
To improve user experience with Spark SQL, and to conform to the SQL standard.

### Does this PR introduce _any_ user-facing change?
No, it just extends existing behavior of casts.

Before:
```sql
spark-sql> SELECT CAST(1.001002BD AS INTERVAL SECOND);
Error in query: cannot resolve 'CAST(1.001002BD AS INTERVAL SECOND)' due to data type mismatch: cannot cast decimal(7,6) to interval second; line 1 pos 7;
```

After:
```
spark-sql> SELECT CAST(1.001002BD AS INTERVAL SECOND);
0 00:00:01.001002000
```

### How was this patch tested?
By running new tests:
```
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z cast.sql"
$ build/sbt "test:testOnly *CastWithAnsiOffSuite"
$ build/sbt "test:testOnly *CastWithAnsiOnSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@HyukjinKwon @cloud-fan @gengliangwang Could you review this PR, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 1]
```diff
@@ -1283,6 +1283,20 @@ object IntervalUtils {
     intToYearMonthInterval(vInt, startField, endField)
   }
 
+  def decimalToYearMonthInterval(
+      d: Decimal, p: Int, s: Int, startField: Byte, endField: Byte): Int = {
+    try {
+      val months = if (endField == YEAR) d.toBigDecimal * MONTHS_PER_YEAR else d.toBigDecimal
+      months.setScale(0, BigDecimal.RoundingMode.HALF_UP).toIntExact
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:1290]: do we have a sql reference doc for cast? we should mention the rounding behavior.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:1290]: We have the doc for cast: https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html#cast but it says that Spark doesn't support cast of intervals to numerics. Let me open a PR and document recent changes in casting of ANSI intervals.

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @cloud-fan for review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:1290]: @cloud-fan I opened the PR https://github.com/apache/spark/pull/37495 which updates the doc.

**Pull request closed without merging**